### ISSUE TITLE
Websocket url issues

### ISSUE THREAD

**Issue by shamblett** [dart-lang/sdk#30523]
The Dart IO WebSocket connect method docs state 

Create a new WebSocket connection. The URL supplied in [url]
   * must use the scheme `ws` or `wss`.

Why is this? It seems restrictive, for example many MQTT brokers support connections over TCP and websockets by just changing the port number, example 

broker.mqttdashboard.com:1883 - TCP
broker.mqttdashboard.com:8080 - Websocket

Could this not be relaxed and allow the user to chose whether to use a scheme or not?

**Comment by lrhn** [comment]
The `WebSocket` class is a client for the web-socket protocol, just as the `HttpClient` is a client for the http protocol. That is why it only accepts URIs that specify a web-socket server, it wouldn't know what to do with anything else.

An MQTT client can connect directly to a broker using the MQTT protocol (over TCP). To specify that with a URI, you should probably use the `mqtt` or `mqtts` scheme, and you would still need an MQTT client implemenetation, who can speak the MQTT protocol natively, to perform that connection.

To connect directly using plain TCP, you can use the `Socket` class.

So, removing the restriction from `WebSocket` that the URI must have a `ws` or `wss` scheme, would not change anything. The class would still only be able to make web-socket connections anyway.

**Comment by shamblett** [comment]
Yes, I have an mqtt_client that does the above, what I'm saying is that using TCP with the Socket class I can just pass a host e.g broker.mqttdashboard.com and it works, to use the same host with the WebSocket class I have to use ws://broker.mqttdashboard.com, its seems odd that one demands a scheme and the other doesnt.

Your right about the mqtt:// schemes, unfortunately in the real world real brokers don't do this, I think also the MQTT spec says a broker operating over websockets should be of the form 

ws://broker.mqttdashboard.com/mqtt

I''v not found a test one yet that does this.

**Comment by lrhn** [comment]
Its true that the `WebSocket` only accepts `Uri` as a way to specify the server, while `Socket` accepts any internet address (as IP address or host name string,  t its own DNS lookup on the name if necessary).

That's probably a symptom of the WebSocket protocol being designed for the web to begin with - even the specification says that each socket has a URL associated with it, so passing in anything but a URL just means that it has to create a URL from that. WebSockets are simply defined, from the ground up, around URIs, so that's what it uses. Pure TCP uses IP addresses and host names, so that's what `Socket` expects, and it doesn't accept a `Uri` at all.

I'll reopen and keep this as a request for having more constuctors on `WebSocket`, but it's not clear that it's worth its own weight since you can just use `new Uri(scheme: "ws", host: host, port: port)` to create the URI (or even `Uri.parse("ws://$host:$port")`).

**Comment by shamblett** [comment]
Yes, this is what I've done, using an independent test client(Hive MQ dashboard) if I open a websocket connection to broker.mqttdashboard.com:8080 it works, if I do it to ws://broker.mqttdashboard.com:8080 the client reports a connection timeout(which is what my mqtt_client reports, so they at least do the same thing).

This looks like a prob with the test broker at the mo, really I'm just looking for a workround to test my mqtt client against a known good broker. I have written a test broker in my test suite which uses web sockets for which I use ws://localhost/mqtt, this works perfectly. 

I'm not saying Dart is wrong, in my experience its usually right as far as standards are concerned, the prob comes when others don't follow the standards.
